### PR TITLE
Added support for VLAN header

### PR DESCRIPTION
**Pull request by gshimansky**
Added functions to parse packets which may have VLAN header. More functions are necessary to generate such packets.

### CODE REVIEW

File: packet/packet.go [hunk 1]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
...
+	} else {
+		packet.L3 = unsafe.Pointer(ptr)
+		return packet.Ether.EtherType, nil
+	}
+}
+
+// GetVLAN returns VLAN header pointer if it is present in the packet.
+func (packet *Packet) GetVLAN() *VLANHdr {
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		return (*VLANHdr)(unsafe.Pointer(uintptr(unsafe.Pointer(packet.Ether)) + EtherLen))
```
Comment (ifilippov) [inline packet/packet.go]: packet.Ether + EtherLen **-2** - no? VLAN tag not after etherHdr, but inside it, before etherType which is 2 byte long

File: packet/packet.go [hunk 2]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
...
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		ptr += VLANLen
+	}
+	packet.L3 = unsafe.Pointer(ptr)
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible VLAN
+// presence into account. It returns VLAN header if present and always
+// correct EtherType from VLAN header or Ethernet header.
+func (packet *Packet) ParseVLANAndL3() (uint16, *VLANHdr) {
```
Comment (ifilippov) [inline packet/packet.go]: I would like to name this function ParseL3ReturnVLAN. Also I suppose that this functions should return only vhdr. User may compare it to to nil after this function.

File: packet/packet.go [hunk 3]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
...
+	ptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		vhdr := (*VLANHdr)(unsafe.Pointer(ptr))
```
Comment (ifilippov) [inline packet/packet.go]: ptr **-2** - no?

File: packet/packet.go [hunk 4]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
 	packet.L3 = unsafe.Pointer(packet.unparsed())
 }
 
+// Correctly returns EtherType from Ethernet header or VLAN header.
```
Comment (ifilippov) [inline packet/packet.go]: Comment should begins with name of function

File: packet/packet.go [hunk 5]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
...
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		vptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+		vhdr := (*VLANHdr)(unsafe.Pointer(vptr))
+		return vhdr.EtherType
+	} else {
+		return packet.Ether.EtherType
+	}
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible presence
```
Comment (ifilippov) [inline packet/packet.go]: Incorrect name of function and word "poinetr"

File: packet/packet.go [hunk 6]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
...
+func (packet *Packet) ParseL3CheckVLAN() {
+	ptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		ptr += VLANLen
+	}
+	packet.L3 = unsafe.Pointer(ptr)
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible VLAN
```
Comment (ifilippov) [inline packet/packet.go]: Incorrect name of function and word "poinetr"

File: packet/packet.go [hunk 7]
```diff
@@ -221,14 +281,34 @@ func (packet *Packet) GetIPv4() *IPv4Hdr {
 	return nil
 }
 
-// GetIPv4 ensures if EtherType is IPv4 and cast L3 poinetr to IPv4Hdr type.
+// GetIPv4CheckVLAN ensures if EtherType is IPv4 and cast L3 poinetr
```
Comment (ifilippov) [inline packet/packet.go]: "poinetr"

File: packet/packet.go [hunk 8]
```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
 	packet.L3 = unsafe.Pointer(packet.unparsed())
 }
 
+// Correctly returns EtherType from Ethernet header or VLAN header.
+func (packet *Packet) GetEtherType() uint16 {
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		vptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
```
Comment (ifilippov) [inline packet/packet.go]: Here and below: we have
"func (packet *Packet) unparsed() uintptr {
	return uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
}"
for this common expression

**Comment by ifilippov** [comment]
Please correct all comments. Also:

ParseL4ForIPv4, ParseL4ForIPv6 use packet.unparsed() and will both be incorrect. They should use "packet.L3" instead of "unparsed()" after this commit. User must parse L3 before L4 - no problems here. If user want - he would use ParseL3CheckVLAN.

Add function ParseAllKnownL3CheckVLAN which will use ParseL3CheckVLAN instead of ParseL3

Please send the result for one more review.

**Comment by ifilippov** [comment]
hm, I understand why you don't use "-2" in ptr arithmetic. You suppose that first 0x8100 is part of etherHdr structure. Technically we can suppose so, however we need to point this clearly and externally. Because standard tells us that 802.1 Q has etherType as first byte and it is always 0x8100. Our structure representation where it is second byte will be very confusing. Please, add appropriate comments to etherHdr structure and VLANHdr structures

**Comment by dkolistr** [comment]
Maybe it also makes sense to make functions for creation of vlan packets.

**Comment by dkolistr** [comment]
Maybe also makes sense to add functions with creation of tagged packets. 
P.S. What if we have double tagging?

File: packet/packet.go [hunk 9]
```diff
@@ -221,14 +281,34 @@ func (packet *Packet) GetIPv4() *IPv4Hdr {
...
+
+// GetARP ensures if EtherType is IPv4 and cast L3 poinetr to IPv4Hdr type.
 func (packet *Packet) GetARP() *ARPHdr {
 	if packet.Ether.EtherType == SwapBytesUint16(ARPNumber) {
 		return (*ARPHdr)(packet.L3)
 	}
 	return nil
 }
 
+// GetARPCheckVLAN ensures if EtherType is IPv4 and cast L3 poinetr to
```
Comment (mbulatova) [inline packet/packet.go]: "GetARPCheckVLAN ensures if EtherType is ARP and cast L3 pointer to ARPHdr type."

Comment (dkolistr) [reply packet/packet.go]: and castS

**Comment by ifilippov** [comment]
Creation - other (next) commit. Double... hm

**Comment by gshimansky** [comment]
I added a comment about our interpretation of VLANHdr. BTW, this is how DPDK interprets it, as an addition for EtherHdr, not a replacement of its part. I removed all unnecessary functions which I don't need, we can refine VLAN API later. So far I left only functionality that is needed for NAT implementation.

**Comment by ifilippov** [comment]
Also:

ParseL4ForIPv4, ParseL4ForIPv6 use packet.unparsed() and will both be incorrect. They should use "packet.L3" instead of "unparsed()" after this commit. User must parse L3 before L4 - no problems here. If user want - he would use ParseL3CheckVLAN.

Add function ParseAllKnownL3CheckVLAN which will use ParseL3CheckVLAN instead of ParseL3

**Comment by gshimansky** [comment]
I removed ParseL4ForIPv4, ParseL4ForIPv6. They are not present.

As for adding new functions, I prefer to refrain from adding many because it is not obvious yet which ones are going to be useful and which aren't. I am adding only those that are going to be used in practice.

**Comment by dkolistr** [comment]
I need to create VLAN-tagged packets in test generator and it is very inconvenient now (as it will be for other users too). 
For example I use InitEmptyIPv4UDPPacket and similar functions for creation, but to make the same IPv4UDP packet with vlan tag I should init empty packet and do all by hand? 
I do not see any good solution instead of duplication a bunch of code making analogous  InitEmptyIPv4UDPPacketVLAN for each init function which adds 4 bytes to size, uses ParseL3CheckVLAN instead of ParseL3 and fills tag (which may come as parameter). But  duplication is always bad...does anyone have ideas how to make it in a better way?

**Comment by ifilippov** [comment]
Daria, You can use our current functions and then EncapsulateHead(0, 4) and then some additional L2 parsing and settinf 0x8100 value. All other protocols will be at their places. So new functions will be {
currentFunction
Encapsulate
Parse
0x8100
}

Let's discuss this further by e-mails.

**Comment by ifilippov** [comment]
Approved

**Pull request merged**